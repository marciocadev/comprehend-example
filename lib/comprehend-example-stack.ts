import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Integration, IntegrationOptions, IntegrationType, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { EventBus, IEventBus, Rule } from 'aws-cdk-lib/aws-events';
import { SfnStateMachine } from 'aws-cdk-lib/aws-events-targets';
import { PolicyDocument, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Code, Function, IFunction, LayerVersion, Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { StateMachine, Wait, WaitTime } from 'aws-cdk-lib/aws-stepfunctions';
import { LambdaInvoke } from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';
import { Action, PolicyStatementFactory } from 'iam-policy-generator';
import { Effect } from 'iam-policy-generator/lib/PolicyFactory';
import { join } from 'path';

export class ComprehendExampleStack extends Stack {

  domainLanguageLambda: IFunction;
  eventBus: IEventBus;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const awsSDKV3Layer = new LayerVersion(this, 'AWSSDKV3', {
      code: Code.fromAsset(join(__dirname, './layer')),
      compatibleRuntimes: [Runtime.NODEJS_14_X]
    });

    const domainLanguageStatement = new PolicyStatement({
      effect: Effect.ALLOW,
      resources: ['*'],
      actions: [
        'comprehend:DetectDominantLanguage'
      ]
    });

    // const domainLanguageStatement = new PolicyStatementFactory()
    //   .setEffect(Effect.ALLOW)
    //   .addResource('*')
    //   .addAction(Action.COMPREHEND.DETECT_DOMINANT_LANGUAGE)
    //   .build();

    this.domainLanguageLambda = new Function(this, 'MyDomainLanguageLambda', {
      //entry: join(__dirname, './lambda-fns/domain-language-lmb/index.ts'),
      code: Code.fromAsset('lib/lambda-fns/domain-language-lmb'),
      handler: 'index.handler',
      runtime: Runtime.NODEJS_14_X,
      layers: [awsSDKV3Layer]
      // bundling: {
      //   minify: true
      // }
    });
    this.domainLanguageLambda.addToRolePolicy(domainLanguageStatement);

    const waitTask = new Wait(this, 'WaitUntil', {
      time: WaitTime.duration(Duration.seconds(10))//.timestampPath('$.detail.at')
    });
    const lambdaTask = new LambdaInvoke(this, 'LambdaInvoke', {
      lambdaFunction: this.domainLanguageLambda,
      outputPath: '$.Payload'
    });

    const stateMachine = new StateMachine(this, 'TextStateMachine', {
      definition: waitTask.next(lambdaTask)
    });

    this.eventBus = new EventBus(this, 'TextEventBus', { 
      eventBusName: 'TextEventBus' 
    });
    new Rule(this, 'LambdaProcessorRule', {
      eventBus: this.eventBus,
      eventPattern: {source:['apigateway'], detailType:['text']},
      targets: [
        new SfnStateMachine(stateMachine)
      ]
    });
    const apigwEventBusRole = new Role(this, 'ApiGatewayEventBusRole', {
      assumedBy: new ServicePrincipal('apigateway'),
      inlinePolicies: {
        putEvents: new PolicyDocument({
          statements: [new PolicyStatement({
            actions: ['events:PutEvents'],
            resources: [this.eventBus.eventBusArn]
          })]
        })
      }
    });

    const options:IntegrationOptions = {
      credentialsRole: apigwEventBusRole,
      requestParameters: {
        "integration.request.header.X-Amz-Target": "'AWSEvents.PutEvents'",
        "integration.request.header.Content-Type": "'application/x-amz-json-1.1'"
      },
      requestTemplates: {
        "application/json": `{"Entries": [
          {"Source": "apigateway", 
          "Detail": "$util.escapeJavaScript($input.body)", 
          "DetailType": "text", 
          "EventBusName": "${this.eventBus.eventBusName}"}]}`
      },
      integrationResponses: [{
        statusCode: "200",
        responseTemplates: {"application/json": ""}
      },
      {
        statusCode: "400",
        responseTemplates: {"application/json": JSON.stringify(
          { state: 'error', message: "$util.escapeJavaScript($input.path('$.errorMessage'))" })}
      }]
    };
    const gw = new RestApi(this, 'TextApiGateway', {deployOptions: {stageName: 'dev'}});
    gw.root.addMethod('POST', new Integration({
        type: IntegrationType.AWS,
        uri: 'arn:aws:apigateway:us-east-1:events:path//',
        integrationHttpMethod: 'POST',
        options:options
      }),
      {
        methodResponses: [{statusCode: "200"}, {statusCode: "400"}]
      }
    );
  }
}
