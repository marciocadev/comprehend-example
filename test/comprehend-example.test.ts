import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { CfnEventBus } from "aws-cdk-lib/aws-events";
import { CfnFunction } from "aws-cdk-lib/aws-lambda";
import { ComprehendExampleStack } from "../lib/comprehend-example-stack";

test('Snapshot', () => {
    // GIVEN
    const app = new App();
    // WHEN
    const stack = new ComprehendExampleStack(app, 'MyTestStack');
    // THEN
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
});

describe('Stack tests', () => {
    let app:App;
    let stack:ComprehendExampleStack;
    let template:Template;

    beforeAll(() => {
        // GIVEN
        app = new App();
        // WHEN
        stack = new ComprehendExampleStack(app, 'MyTestStack');
        // THEN
        template = Template.fromStack(stack);
    });

    test('EventBus Created', () => {
        template.findResources('AWS::Events::EventBus');
        template.hasResourceProperties('AWS::Events::EventBus', {
            Name: 'TextEventBus'
        })
    });

    test('Event Rule Created', () => {
        const lambdaLogicalId = stack.resolve((stack.domainLanguageLambda.node.defaultChild as CfnFunction).logicalId)
        console.log(lambdaLogicalId);

        const eventBusLogicalId = stack.resolve((stack.eventBus.node.defaultChild as CfnEventBus).logicalId);

        template.resourceCountIs('AWS::Events::Rule', 1);
        template.findResources('AWS::Events::Rule');
        template.hasResourceProperties('AWS::Events::Rule', {
            State: 'ENABLED',
            EventPattern: {
                source: ['apigateway'],
                'detail-type': ['text']
            },
            EventBusName: {
                Ref: eventBusLogicalId
            }
            // Targets: [{
            //     Arn: {
            //         'Fn::GetAtt': [
            //             'LambdaD247545B',
            //             'Arn'
            //         ]
            //     },
            //     Id: 'Target0'
            // }]
        });
    });

    test('RespApi Created', () => {
        template.findResources('AWS::ApiGateway::RestApi');
        template.hasResourceProperties('AWS::ApiGateway::RestApi', {
            Name: 'TextApiGateway'
        });
    });
});
