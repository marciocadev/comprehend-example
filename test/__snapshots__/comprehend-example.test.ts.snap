// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "TextApiGatewayEndpoint1B713D84": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "TextApiGateway437DBC5F",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "TextApiGatewayDeploymentStagedev47E27826",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AWSSDKV3FE048A04": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "nodejs14.x",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "df49ab4797665962d0e5700f2b5cb624d01a72f0ed92ab60135ac2d58a2977c1.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ApiGatewayEventBusRole78CD632E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "events:PutEvents",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "TextEventBus50F77630",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "putEvents",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaProcessorRuleC28EF8EB": Object {
      "Properties": Object {
        "EventBusName": Object {
          "Ref": "TextEventBus50F77630",
        },
        "EventPattern": Object {
          "detail-type": Array [
            "text",
          ],
          "source": Array [
            "apigateway",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "TextStateMachine95BF0DB2",
            },
            "Id": "Target0",
            "RoleArn": Object {
              "Fn::GetAtt": Array [
                "TextStateMachineEventsRole7E52DA46",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MyDomainLanguageLambda661AC446": Object {
      "DependsOn": Array [
        "MyDomainLanguageLambdaServiceRoleDefaultPolicyE5218BC0",
        "MyDomainLanguageLambdaServiceRole21F2C121",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "79f25cd514ff983e765920a2e8a5a22b0277ac59c6e6cc4d1bda7ccc3946378c.zip",
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "AWSSDKV3FE048A04",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyDomainLanguageLambdaServiceRole21F2C121",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyDomainLanguageLambdaServiceRole21F2C121": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyDomainLanguageLambdaServiceRoleDefaultPolicyE5218BC0": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "comprehend:DetectDominantLanguage",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyDomainLanguageLambdaServiceRoleDefaultPolicyE5218BC0",
        "Roles": Array [
          Object {
            "Ref": "MyDomainLanguageLambdaServiceRole21F2C121",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TextApiGateway437DBC5F": Object {
      "Properties": Object {
        "Name": "TextApiGateway",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "TextApiGatewayAccount54FA72E1": Object {
      "DependsOn": Array [
        "TextApiGateway437DBC5F",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "TextApiGatewayCloudWatchRoleE5030C8C",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "TextApiGatewayCloudWatchRoleE5030C8C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TextApiGatewayDeployment754CD196c3c1adbc4ea08c0597ce4b42e4ae4328": Object {
      "DependsOn": Array [
        "TextApiGatewayPOST0201D577",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "TextApiGateway437DBC5F",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "TextApiGatewayDeploymentStagedev47E27826": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "TextApiGatewayDeployment754CD196c3c1adbc4ea08c0597ce4b42e4ae4328",
        },
        "RestApiId": Object {
          "Ref": "TextApiGateway437DBC5F",
        },
        "StageName": "dev",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "TextApiGatewayPOST0201D577": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "ApiGatewayEventBusRole78CD632E",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "ResponseTemplates": Object {
                "application/json": "",
              },
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "{\\"state\\":\\"error\\",\\"message\\":\\"$util.escapeJavaScript($input.path('$.errorMessage'))\\"}",
              },
              "StatusCode": "400",
            },
          ],
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'application/x-amz-json-1.1'",
            "integration.request.header.X-Amz-Target": "'AWSEvents.PutEvents'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{\\"Entries\\": [
          {\\"Source\\": \\"apigateway\\", 
          \\"Detail\\": \\"$util.escapeJavaScript($input.body)\\", 
          \\"DetailType\\": \\"text\\", 
          \\"EventBusName\\": \\"",
                  Object {
                    "Ref": "TextEventBus50F77630",
                  },
                  "\\"}]}",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": "arn:aws:apigateway:us-east-1:events:path//",
        },
        "MethodResponses": Array [
          Object {
            "StatusCode": "200",
          },
          Object {
            "StatusCode": "400",
          },
        ],
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "TextApiGateway437DBC5F",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "TextApiGateway437DBC5F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TextEventBus50F77630": Object {
      "Properties": Object {
        "Name": "TextEventBus",
      },
      "Type": "AWS::Events::EventBus",
    },
    "TextStateMachine95BF0DB2": Object {
      "DependsOn": Array [
        "TextStateMachineRoleDefaultPolicy7A727CB5",
        "TextStateMachineRoleA1896FF6",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"WaitUntil\\",\\"States\\":{\\"WaitUntil\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":10,\\"Next\\":\\"LambdaInvoke\\"},\\"LambdaInvoke\\":{\\"End\\":true,\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"OutputPath\\":\\"$.Payload\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "MyDomainLanguageLambda661AC446",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "TextStateMachineRoleA1896FF6",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "TextStateMachineEventsRole7E52DA46": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TextStateMachineEventsRoleDefaultPolicyC69D252C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "TextStateMachine95BF0DB2",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TextStateMachineEventsRoleDefaultPolicyC69D252C",
        "Roles": Array [
          Object {
            "Ref": "TextStateMachineEventsRole7E52DA46",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TextStateMachineRoleA1896FF6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TextStateMachineRoleDefaultPolicy7A727CB5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "MyDomainLanguageLambda661AC446",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TextStateMachineRoleDefaultPolicy7A727CB5",
        "Roles": Array [
          Object {
            "Ref": "TextStateMachineRoleA1896FF6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
